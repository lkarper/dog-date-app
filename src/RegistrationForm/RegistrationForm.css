.RegistrationForm__header h1 {
    text-align: center;
}

.RegistrationForm__loading-container {
    width: 100%;
    height: 100%;
    position: fixed;
    top: 75px;
    left: 0;
    background-image: linear-gradient(
        rgba(0, 0, 0, 0.5),
        rgba(0, 0, 0, 0.5)
    );
    display: flex;
    justify-content: center;
    align-items: center;
}

.RegistrationForm__loading {
    font-size: 30px;
}