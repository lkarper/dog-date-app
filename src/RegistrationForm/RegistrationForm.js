import React, { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import ValidateEmail from '../validation-components/ValidateEmail';
import ValidatePhoneNumber from '../validation-components/ValidatePhoneNumber';
import ValidateUsername from '../validation-components/ValidateUsername';
import ValidatePassword from '../validation-components/ValidatePassword';
import ValidateReenteredPassword from '../validation-components/ValidateReenteredPassword';
import STORE from '../STORE';

const RegistrationForm = (props) => {

    const [email, setEmail] = useState('');
    const [emailValidationError, setEmailValidationError] = useState(null);
    const [phone, setPhone] = useState('');
    const [phoneValidationError, setPhoneValidationError] = useState(null);
    const [username, setUsername] = useState('');
    const [usernameValidationError, setUsernameValidationError] = useState(null);
    const [password, setPassword] = useState('');
    const [passwordErrorMessage, setPasswordErrorMessage] = useState([]);
    const [passwordError, setPasswordError] = useState({
        tooShort: true,
        tooLong: false,
        endSpaces: false,
        upperCase: true,
        lowerCase: true,
        number: true,
        specialChar: true,
    });
    const [reenteredPassword, setReenteredPassword] = useState('');
    const [reenteredPasswordError, setReenteredPasswordError] = useState(null);
    const [submissionError, setSubmissionError] = useState(null);
    const [emailAlreadyRegistered, setEmailAlreayRegistered] = useState(false);
    const [usernameExists, setUsernameExists] = useState(false);

    const handleSubmit = (event) => {
        event.preventDefault();
        if (STORE.users.find(user => user.email === email)) {
            setEmailAlreayRegistered(true);
        }
        if (STORE.users.find(user => user.username === username)) {
            setUsernameExists(true);
        }
        if (!emailAlreadyRegistered && ! usernameExists) {
            const newUser = {
                id: uuidv4(),
                email,
                username,
                phone,
                password
            }
            STORE.users.push(newUser);
            props.history.push('/login');
        }
    }

    return (
        <section>
            <header>
                <h2>Even man's best friend needs a pack.</h2>
                <p>Create an account to get started!</p>
            </header>
            <form 
                className='signup-form'
                onSubmit={(e) => handleSubmit(e)}
            >
                <div>
                    <label htmlFor="email">Email:</label>
                    <input 
                        type="email" 
                        placeholder="example@example.com" 
                        id="email" 
                        name="email"
                        value={email} 
                        onChange={(e) => setEmail(e.target.value)}
                        required
                        aria-describedby='email-validation'
                    />
                </div>
                <div role='alert'>
                    <ValidateEmail 
                        email={email}
                        emailValidationError={emailValidationError}
                        setEmailValidationError={setEmailValidationError}
                    />
                </div>
                <div>
                    <label htmlFor="phone">Phone number:</label>
                    <input 
                        type="tel" 
                        id="phone" 
                        name="phone"
                        value={phone} 
                        pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}" 
                        aria-describedby="phone-validation" 
                        onChange={(e) => setPhone(e.target.value)}
                    />
                </div>
                <div role='alert'>
                    <ValidatePhoneNumber 
                        phone={phone}
                        setPhoneValidationError={setPhoneValidationError}
                    /> 
                </div>
                <div>
                    <label htmlFor="username">Username:</label>
                    <input 
                        type="text" 
                        placeholder="Username" 
                        id="username" 
                        name="username"
                        value={username} 
                        aria-describedby='username-validation'
                        onChange={(e) => setUsername(e.target.value)}
                    />
                </div>
                <div role='alert'>
                    <ValidateUsername 
                        username={username}
                        usernameValidationError={usernameValidationError}
                        setUsernameValidationError={setUsernameValidationError}
                    />
                </div>
                <div>
                    <label htmlFor="password">Password:</label>
                    <input 
                        type="password" 
                        id="password" 
                        name="password" 
                        value={password}
                        aria-describedby="password-error-message"
                        onChange={(e) => setPassword(e.target.value)}
                        required
                        />
                </div>
                <ValidatePassword 
                    password={password} 
                    passwordErrorMessage={passwordErrorMessage}
                    passwordError={passwordError}
                    setPasswordErrorMessage={setPasswordErrorMessage}
                    setPasswordError={setPasswordError}
                />
                <div>
                    <label htmlFor="reenter-password">Re-enter password:</label>
                    <input 
                        type="password" 
                        id="reenter-password" 
                        name="reenter-password" 
                        value={reenteredPassword}
                        aria-describedby="reenter-password-validation"
                        onChange={(e) => setReenteredPassword(e.target.value)}
                        required
                    />
                </div>
                <ValidateReenteredPassword 
                    password={password}
                    reenteredPassword={reenteredPassword}
                    reenteredPasswordError={reenteredPasswordError}
                    setReenteredPasswordError={setReenteredPasswordError}
                /> 
                <div role='alert'>

                </div>
                <button 
                    type="submit"
                    disabled={
                        emailValidationError || 
                        phoneValidationError || 
                        usernameValidationError || 
                        passwordErrorMessage.length || 
                        reenteredPasswordError}
                >
                    Submit
                </button>
            </form>
        </section>
    );
}

export default RegistrationForm;