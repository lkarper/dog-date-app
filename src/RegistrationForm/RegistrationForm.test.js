import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import RegistrationForm from './RegistrationForm';

describe('RegistrationForm component', () => {
    it('renders the UI without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <BrowserRouter>
                <RegistrationForm />
            </BrowserRouter>, 
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const wrapper = mount(
            <BrowserRouter>
                <RegistrationForm />
            </BrowserRouter>
        );
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });
});
