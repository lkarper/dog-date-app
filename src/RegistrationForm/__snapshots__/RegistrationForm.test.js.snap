// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm component renders the UI as expected 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <RegistrationForm
      currentEmail=""
      currentPhone=""
      history={
        Object {
          "push": [Function],
        }
      }
      setShowEdit={[Function]}
      suffix=""
    >
      <section
        className="RegistrationForm__section outer-section"
      >
        <header
          className="RegistrationForm__header"
        >
          <h1>
            Even man's best friend needs a pack...
          </h1>
        </header>
        <form
          className="RegistrationForm__form"
          onSubmit={[Function]}
        >
          <fieldset
            className="RegistrationForm__fieldset outer-fieldset "
          >
            <legend>
              Create an account to get started!
            </legend>
            <div>
              <label
                htmlFor="email"
              >
                Email:
              </label>
              <input
                aria-describedby="email-validation"
                id="email"
                name="email"
                onChange={[Function]}
                placeholder="example@example.com"
                required={true}
                type="email"
                value=""
              />
            </div>
            <div
              role="alert"
            >
              <ValidateEmail
                email=""
                emailAlreadyRegistered=""
                emailValidationError="Email invalid."
                setEmailValidationError={[Function]}
                suffix=""
              >
                <p
                  className="ValidateEmail__validator error"
                  id="email-validation"
                >
                  A valid email address is required.
                  <br />
                </p>
              </ValidateEmail>
            </div>
            <div>
              <label
                htmlFor="phone"
              >
                Phone number:
              </label>
              <input
                aria-describedby="phone-validation"
                id="phone"
                name="phone"
                onChange={[Function]}
                pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
                type="tel"
                value=""
              />
            </div>
            <div
              role="alert"
            >
              <ValidatePhoneNumber
                phone=""
                phoneValidationError=""
                setPhoneValidationError={[Function]}
              >
                <p
                  className="ValidatePhoneNumber__validator  "
                  id="phone-validation"
                >
                  Phone number optional, but if provided must match format '123-456-7890'.
                </p>
              </ValidatePhoneNumber>
            </div>
            <div>
              <label
                htmlFor="username"
              >
                Username:
              </label>
              <input
                aria-describedby="username-validation"
                autoComplete="username"
                id="username"
                name="username"
                onChange={[Function]}
                placeholder="Username"
                type="text"
                value=""
              />
            </div>
            <div
              role="alert"
            >
              <ValidateUsername
                setUsernameValidationError={[Function]}
                username=""
                usernameExists=""
                usernameValidationError="Username required and must be 3-72 characters in length."
              >
                <p
                  className="ValidateUserName__validator error"
                  id="username-validation"
                >
                  Username required and must be 3-72 characters in length.
                  <br />
                </p>
              </ValidateUsername>
            </div>
            <div>
              <label
                htmlFor="password"
              >
                Password:
              </label>
              <input
                aria-describedby="password-error-message"
                autoComplete="new-password"
                id="password"
                name="password"
                onChange={[Function]}
                required={true}
                type="password"
                value=""
              />
            </div>
            <ValidatePassword
              password=""
              passwordError={
                Object {
                  "endSpaces": false,
                  "lowerCase": true,
                  "number": true,
                  "specialChar": true,
                  "tooLong": false,
                  "tooShort": true,
                  "upperCase": true,
                }
              }
              passwordErrorMessage={
                Array [
                  "Password must be at least 8 characters in length.",
                  "Password must contain at least one lowercase character.",
                  "Password must contain at least one uppercase character.",
                  "Password must contain at least one number.",
                  "Password must contain at least one special character.",
                ]
              }
              setPasswordError={[Function]}
              setPasswordErrorMessage={[Function]}
            >
              <PasswordChecklist
                passwordError={
                  Object {
                    "endSpaces": false,
                    "lowerCase": true,
                    "number": true,
                    "specialChar": true,
                    "tooLong": false,
                    "tooShort": true,
                    "upperCase": true,
                  }
                }
              >
                <div>
                  <p>
                    Password must:
                  </p>
                  <ul>
                    <li>
                      Be at least 8 characters in length 
                      <span
                        style={
                          Object {
                            "color": "red",
                          }
                        }
                      >
                        ✖
                      </span>
                    </li>
                    <li>
                      Be no more than 72 characters in length 
                      <span
                        style={
                          Object {
                            "color": "hsl(245, 100%, 50%)",
                          }
                        }
                      >
                        ✔
                      </span>
                    </li>
                    <li>
                      Not begin or end with a space 
                      <span
                        style={
                          Object {
                            "color": "hsl(245, 100%, 50%)",
                          }
                        }
                      >
                        ✔
                      </span>
                    </li>
                    <li>
                      Contain at least one lowercase letter 
                      <span
                        style={
                          Object {
                            "color": "red",
                          }
                        }
                      >
                        ✖
                      </span>
                    </li>
                    <li>
                      Contain at least one uppercase letter 
                      <span
                        style={
                          Object {
                            "color": "red",
                          }
                        }
                      >
                        ✖
                      </span>
                    </li>
                    <li>
                      Contain at least one number 
                      <span
                        style={
                          Object {
                            "color": "red",
                          }
                        }
                      >
                        ✖
                      </span>
                    </li>
                    <li>
                      Contain at least one special character (e.g. !, @, #, $, etc.) 
                      <span
                        style={
                          Object {
                            "color": "red",
                          }
                        }
                      >
                        ✖
                      </span>
                    </li>
                  </ul>
                </div>
              </PasswordChecklist>
              <div
                className="ValidatePassword__password-error-message"
                role="alert"
              >
                <p
                  id="password-error-message"
                >
                  Password must be at least 8 characters in length. Password must contain at least one lowercase character. Password must contain at least one uppercase character. Password must contain at least one number. Password must contain at least one special character.
                </p>
              </div>
            </ValidatePassword>
            <div>
              <label
                htmlFor="reenter-password"
              >
                Re-enter password:
              </label>
              <input
                aria-describedby="reenter-password-validation"
                autoComplete="new-password"
                id="reenter-password"
                name="reenter-password"
                onChange={[Function]}
                required={true}
                type="password"
                value=""
              />
            </div>
            <ValidateReenteredPassword
              password=""
              reenteredPassword=""
              reenteredPasswordError=""
              setReenteredPasswordError={[Function]}
            >
              <p
                className="ValidateReenteredPassword__validator valid"
                id="reenter-password-validation"
              >
                Passwords match.
              </p>
            </ValidateReenteredPassword>
          </fieldset>
          <button
            className="button"
            disabled="Email invalid."
            type="submit"
          >
            Submit
          </button>
        </form>
        <div
          className="RegistrationForm__alert-div"
          role="alert"
        />
      </section>
    </RegistrationForm>
  </Router>
</BrowserRouter>
`;
